Get rid of dim_groups attribute from DeviceMesh

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #103106
* __->__ #103105

This PR get rids of the dim_groups attribute from DeviceMesh, the main
motivation behind this is that we should let c10d store the process
groups during its creation instead of DeviceMesh, DeviceMesh should just
handle ranks correctly.

This could enable DTensor becomes picklable! (torch.save/load could be
possible), which I will give it a try in the next PR

wanchaol committed:
Get rid of dim_groups attribute from DeviceMesh

This PR get rids of the dim_groups attribute from DeviceMesh, the main
motivation behind this is that we should let c10d store the process
groups during its creation instead of DeviceMesh, DeviceMesh should just
handle ranks correctly.

This could enable DTensor becomes picklable! (torch.save/load could be
possible), which I will give it a try in the next PR

[ghstack-poisoned]

wanchaol committed:
Update on "Get rid of dim_groups attribute from DeviceMesh"


This PR get rids of the dim_groups attribute from DeviceMesh, the main
motivation behind this is that we should let c10d store the process
groups during its creation instead of DeviceMesh, DeviceMesh should just
handle ranks correctly.

This could enable DTensor becomes picklable! (torch.save/load could be
possible), which I will give it a try in the next PR

[ghstack-poisoned]

wanchaol committed:
Update on "Get rid of dim_groups attribute from DeviceMesh"


This PR get rids of the dim_groups attribute from DeviceMesh, the main
motivation behind this is that we should let c10d store the process
groups during its creation instead of DeviceMesh, DeviceMesh should just
handle ranks correctly.

This could enable DTensor becomes picklable! (torch.save/load could be
possible), which I will give it a try in the next PR

[ghstack-poisoned]

wanchaol committed:
Update on "Get rid of dim_groups attribute from DeviceMesh"


This PR get rids of the dim_groups attribute from DeviceMesh, the main
motivation behind this is that we should let c10d store the process
groups during its creation instead of DeviceMesh, DeviceMesh should just
handle ranks correctly.

This could enable DTensor becomes picklable! (torch.save/load could be
possible), which I will give it a try in the next PR

[ghstack-poisoned]

XilunWu commented on torch/distributed/_tensor/device_mesh.py:
```diff
@@ -169,15 +170,17 @@ def _init_process_groups(self):
-        dim_groups: List[ProcessGroup] = []
+        # group ranks/tags associated with each mesh dimension, each mesh dimension should
+        # have one sub-group per rank
+        dim_group_ranks: List[List[int]] = []
+        dim_group_tags: List[str] = []
```
suggestion: 
since we're mostly (if not always) using tag and list of ranks at the same time, shall we combine (zip together) them into a single `List[Tuple[str, List[int]]]` where each element in `List` is a `Tuple` of a tag and a subgroup?

wanchaol replied:
sure, updated

XilunWu commented on torch/distributed/_tensor/device_mesh.py:
```diff
@@ -169,15 +170,17 @@ def _init_process_groups(self):
             # pg to the first dim groups, as new_group cannot have the exact
             # same ranks as world
-            dim_groups.append(default_pg)
+            dim_group_ranks.append(list(range(get_world_size())))
+            dim_group_tags.append(_get_group_tag(_get_default_group()))
```
suggestion:
then we can do something like:
`dim_groups.append((tag, list(range(world_size))))`

XilunWu commented on torch/distributed/_tensor/device_mesh.py:
```diff
@@ -193,16 +196,18 @@ def _init_process_groups(self):
                             )
-                        dim_groups.append(new_subgroup)
-        return dim_groups
+                        dim_group_ranks.append(subgroup_ranks)
+                        dim_group_tags.append(_get_group_tag(dim_group))
+        self._dim_group_ranks = dim_group_ranks
+        self._dim_group_tags = dim_group_tags
```
same as above

XilunWu commented on torch/distributed/_tensor/device_mesh.py:
```diff
@@ -227,10 +232,24 @@ def __eq__(self, other: object) -> bool:
 
-    def get_dim_groups(self) -> List[ProcessGroup]:
-        if not hasattr(self, "_dim_groups"):
+    def get_dim_groups(
+        self, mesh_dim: Optional[int] = None
+    ) -> Union[ProcessGroup, List[ProcessGroup]]:
```
Will we need to add a `instanceOf` after existing `get_dim_groups` calls to make `pyre` happy?

wanchaol replied:
I added assert isinstance so that mypy happy

XilunWu commented on torch/distributed/_tensor/device_mesh.py:
```diff
@@ -227,10 +232,24 @@ def __eq__(self, other: object) -> bool:
+    def get_dim_groups(
+        self, mesh_dim: Optional[int] = None
+    ) -> Union[ProcessGroup, List[ProcessGroup]]:
+        if not hasattr(self, "_dim_group_ranks"):
```
same as above, if we combine ranks and tag into a tuple, we just use "_dim_groups" here.

XilunWu commented on torch/distributed/_tensor/device_mesh.py:
```diff
@@ -227,10 +232,24 @@ def __eq__(self, other: object) -> bool:
-        return self._dim_groups
+        if mesh_dim is not None:
+            return _find_pg_by_ranks_and_tag(
+                self._dim_group_tags[mesh_dim], self._dim_group_ranks[mesh_dim]
```
`self._dim_group_tags[mesh_dim], self._dim_group_ranks[mesh_dim]` -> `*self._dim_groups[mesh_dim]`

XilunWu commented on torch/distributed/_tensor/device_mesh.py:
```diff
@@ -227,10 +232,24 @@ def __eq__(self, other: object) -> bool:
+            for mesh_dim in range(self.mesh.ndim):
+                dim_groups.append(
+                    _find_pg_by_ranks_and_tag(
+                        self._dim_group_tags[mesh_dim], self._dim_group_ranks[mesh_dim]
```
same as above

XilunWu commented on torch/distributed/_tensor/device_mesh.py:
```diff
@@ -279,9 +298,11 @@ def scatter(
             return None
-        dim_group = self._dim_groups[mesh_dim]
+        dim_group = self.get_dim_groups(mesh_dim)
+        assert isinstance(dim_group, ProcessGroup)
```
yeah so it seems we need to do this everywhere after calling `device_mesh.get_dim_groups()`. We probably should avoid this.

wanchaol replied:
`get_dim_groups` now return either a ProcessGroup or a List of ProcessGroup depending on whether you pass a mesh dim or not, this assert is simply a type check to make sure mypy does not complain

XilunWu replied:
Yes I know. But I feel that having to add an assertion every time after calling `get_dim_groups` is not a good sign and we should avoid that by changing the return type.  Is there a reason why we want to return a List of `ProcessGroup`s instead of one PG that contains all the ranks within the list?

wanchaol replied:
i don't understand what you mean by "changing the return type". Can you clarify more on this? The reason why we want to return a list of process groups is that user want to have those groups for different mesh dimension, i.e. the 2-D parallel need this API to work with FSDP.

What do you mean by "instead of one PG that contains all the ranks within the list"?

XilunWu replied:
I propose that we separate this function into 2. One takes in a `dim` argument and returns a `ProcessGroup` (`get_group_by_dim`) and the other simply returns a `List[ProcessGroup]` (`get_dim_groups`), because adding assertion is quite not ideal.

With regard to "instead of one PG that contains all the ranks within the list", forget it. My mistake, I misunderstood the function's behavior.

XilunWu commented on torch/distributed/tensor/parallel/_utils.py:
```diff
@@ -151,6 +151,6 @@ def _create_1d_device_mesh(device_mesh: DeviceMesh, tp_mesh_dim: int = 0) -> Dev
 
-    sub_pg = device_mesh.get_dim_groups()[tp_mesh_dim]
-    res_sub_mesh._dim_groups = [sub_pg]
+    res_sub_mesh._dim_group_ranks = [device_mesh._dim_group_ranks[tp_mesh_dim]]
+    res_sub_mesh._dim_group_tags = [device_mesh._dim_group_tags[tp_mesh_dim]]
```
we can combine them into one line with the suggestion on top, right? I think that would be better, what do you think?

XilunWu commented on torch/distributed/tensor/parallel/fsdp.py:
```diff
@@ -184,7 +184,9 @@ def _create_sharded_tensor_md_from_dt(
     assert mesh.ndim == 1, "Only 1D DeviceMeshes currently handled"
-    return mesh.get_dim_groups()[0]
+    dim_groups = mesh.get_dim_groups()
+    assert isinstance(dim_groups, List[c10d.ProcessGroup])
```
`assert isinstance(..., List)` but we better get rid of checking types.

wanchaol replied:
there's no way to get rid of them if making `get_dim_groups` return two different types, and I think type check is a good thing for python program to make sure there's no apparent bugs

XilunWu reviewed (commented):
Yeah we start to make DeviceMesh pickable, thanks for the work!! Some suggestions (I could be wrong). Let me know what you think ;-)

wanchaol committed:
Update on "Get rid of dim_groups attribute from DeviceMesh"


This PR get rids of the dim_groups attribute from DeviceMesh, the main
motivation behind this is that we should let c10d store the process
groups during its creation instead of DeviceMesh, DeviceMesh should just
handle ranks correctly.

This could enable DTensor becomes picklable! (torch.save/load could be
possible), which I will give it a try in the next PR

[ghstack-poisoned]

wanchaol committed:
Update on "Get rid of dim_groups attribute from DeviceMesh"


This PR get rids of the dim_groups attribute from DeviceMesh, the main
motivation behind this is that we should let c10d store the process
groups during its creation instead of DeviceMesh, DeviceMesh should just
handle ranks correctly.

This could enable DTensor becomes picklable! (torch.save/load could be
possible), which I will give it a try in the next PR

[ghstack-poisoned]

fduwjj reviewed (commented):
So basically if we store PG as attribute it will make it unpickable but now what we store is `List[Tuple[str, List[int]]]` so it works now. Is this understanding correct?

XilunWu commented:
@fduwjj That is correct. For some reason `ProcessGroup` is not an picklable object.

wanchaol committed:
Update on "Get rid of dim_groups attribute from DeviceMesh"


This PR get rids of the dim_groups attribute from DeviceMesh, the main
motivation behind this is that we should let c10d store the process
groups during its creation instead of DeviceMesh, DeviceMesh should just
handle ranks correctly.

This could enable DTensor becomes picklable! (torch.save/load could be
possible), which I will give it a try in the next PR

[ghstack-poisoned]

XilunWu reviewed (approved):
LGTM!

fduwjj reviewed (approved):
LGTM, hopefully this will not add too much perf overhead.

wanchaol commented:
> LGTM, hopefully this will not add too much perf overhead.

hmmm what do you mean by perf overhead here? I think this would not have any perf implications?

wanchaol committed:
Update on "Get rid of dim_groups attribute from DeviceMesh"


This PR get rids of the dim_groups attribute from DeviceMesh, the main
motivation behind this is that we should let c10d store the process
groups during its creation instead of DeviceMesh, DeviceMesh should just
handle ranks correctly.

This could enable DTensor becomes picklable! (torch.save/load could be
possible), which I will give it a try in the next PR

[ghstack-poisoned]
